source/custom/visa_service.custom.cpp

```diff
@@ -255,6 +257,19 @@ ::grpc::Status VisaService::Open(::grpc::ServerContext* context, const OpenReque
     if (!status_ok(status)) {
       return ConvertApiErrorStatusForViSession(context, status, rsrc_manager_handle);
     }
+
+    ViUInt16 numEvents = 0;
+    ViSession vi = session_repository_->access_session(grpc_device_session_name);
+    if (library->GetAttribute(vi, VI_ATTR_NUM_SUP_EVENTS, &numEvents) == VI_SUCCESS && numEvents > 0) {
+      // NI-VISA writes the events followed by an extra "-1" value that we don't need.
+      EventType* events = new EventType[numEvents + 1];
```

Comment:
std::vector<EventType> events(numEvents+1);
library->GetAttribute(vi, VI_ATTR_SUP_EVENTS, events.data());
for (ViUInt16 i = 0; i < numEvents; ++i) {
   response->add_supported_events(events[i]);
}
// no need to delete it explicitly

Reply:
Of course, much better C++ code. Thanks.